src/inmanta/execute/scheduler.py

```diff
@@ -430,3 +470,220 @@ def run(self, compiler: "Compiler", statements: Sequence["Statement"], blocks: S
+
+    def __init__(self, type_hints: List["TypeHint"]) -> None:
+        self._all_constraint_entity_relationships: Set[EntityRelationship] = set(
+            itertools.chain.from_iterable(EntityRelationship.from_type_hint(hint) for hint in type_hints)
+        )
+
+        self._unconstraint_variables: Deque[DelayedResultVariable[Any]] = deque()
+        self._constraint_variables: Dict[EntityRelationship, Deque[DelayedResultVariable[Any]]] = {
+            entity_relationship: deque() for entity_relationship in self._all_constraint_entity_relationships
+        }
+        self._tmp_list_variables: Deque[TempListVariable] = deque()
+
+        # A queue that indicates a valid order in which the self._constraint_variables have to be returned
+        # This queue is never modified.
+        self._freeze_order: Deque[EntityRelationship] = deque(TypePrecedenceGraph(type_hints).get_freeze_order())
+        # Copy of self._freeze_order. At all times the first element of this queue
+        # points to the next type that should be returned from self._constraint_variables
+        self._freeze_order_working_list: Deque[EntityRelationship] = self._freeze_order.copy()
+
+    def __len__(self) -> int:
+        """
+        Return the number of elements present in this queue.
+        """
+        return (
+            len(self._unconstraint_variables)
+            + sum(len(v) for v in self._constraint_variables.values())
+            + len(self._tmp_list_variables)
+        )
+
+    def append(self, drv: DelayedResultVariable[Any], dont_reset_working_list: bool = False) -> None:
+        """
+        Append on the right side of the queue.
+        """
+        if isinstance(drv, TempListVariable):
+            self._tmp_list_variables.append(drv)
+            return
+        entity_relationship = EntityRelationship.from_delayed_result_variable(drv)
+        if entity_relationship in self._constraint_variables:
+            self._constraint_variables[entity_relationship].append(drv)
+            if not dont_reset_working_list and entity_relationship not in self._freeze_order_working_list:
+                # working list is dirty
+                self._freeze_order_working_list = self._freeze_order.copy()
+        else:
+            self._unconstraint_variables.append(drv)
+
+    def popleft(self) -> DelayedResultVariable[Any]:
+        """
+        Remove element from the left side of the queue and return it.
+        """
+        try:
+            return self._unconstraint_variables.popleft()
+        except IndexError:
+            # Empty
+            pass
+        try:
+            return self._get_next_constraint_variable()
+        except IndexError:
+            # Empty
+            pass
+        return self._tmp_list_variables.popleft()
+
+    def _get_next_constraint_variable(self) -> DelayedResultVariable[Any]:
+        if not self._constraint_variables:
+            raise IndexError()
+        while len(self._freeze_order_working_list) > 0:
```

Comment:
isn't using length slow compared to truthiness? (this is on the compiler critical path)